Library/OcCryptoLib/X64/sha512-avx.nasm

```diff
@@ -0,0 +1,532 @@
+  vpsllq  xmm2, xmm4, (64-8)  ; xmm2 = W[t-15] << 64-8
+  vpor   xmm3, xmm3, xmm2     ; xmm3 = (W[t-15] >> 8) | (W[t-15] << 64-8)
+  vpxor	 xmm1, xmm1, xmm3     ; xmm1 = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8)
+  ; W[t-15] SHR 7
+  vpsrlq  xmm3, xmm4, 7       ; xmm3 = W[t-15] >> 7
+  vpxor	 xmm1, xmm1, xmm3     ; xmm1 = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8) ^ (W[t-15] SHR 7)
+
+  ; W[t] = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+  %assign idx  (%1 - 7)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7]
+  vpaddq xmm0, xmm0, xmm1     ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15])
+  %assign idx  (%1 - 16)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+	%assign idx  %1
+	vmovdqa	W_t(idx), xmm0	    ; Store W[t]
+	vpaddq xmm0, xmm0, K_t(idx) ; Compute W[t]+K[t]
+	vmovdqa	WK_2(idx), xmm0     ; Store W[t]+K[t] for next rounds
+%endmacro
+
+; #######################################################################
+; BOOLEAN IsAvxSupported ()
+; To run in QEMU use options: -enable-kvm -cpu Penryn,+avx,+xsave,+xsaveopt
+; #######################################################################
+align 8
+global ASM_PFX(IsAvxSupported)
+ASM_PFX(IsAvxSupported):
+  ; Detect CPUID.1:ECX.XSAVE[bit 26] = 1 (CR4.OSXSAVE can be set to 1).
+  ; Detect CPUID.1:ECX.AVX[bit 28] = 1 (AVX instructions supported).
+  mov eax, 1          ; Feature Information
+  cpuid               ; result in EAX, EBX, ECX, EDX
+  and ecx, 014000000H
+  cmp ecx, 014000000H ; check both XSAVE and AVX feature flags
+  jne noAVX
+  ; processor supports AVX instructions
+  mov rax, cr4
+  bts rax, 18  ; OSXSAVE: enables XGETBV and XSETBV
+  mov cr4, rax
+
+  mov ecx, 0          ; read the contents of XCR0 register
+  XGETBV              ; result in EDX:EAX
+  or  eax, 06H        ; enable both XMM and YMM state support
+  ; XSETBV must be executed at privilege level 0 or in real-address mode.
+  XSETBV
+  mov rax, 1
+  jmp done
+noAVX:
+  mov rax, 0
+done:
+  ret
+
+; #######################################################################
+;  void Sha512TransformAvx(sha512_state *state, const u8 *data, int blocks)
+;  Purpose: Updates the SHA512 digest stored at "state" with the message
+;  stored in "data".
+;  The size of the message pointed to by "data" must be an integer multiple
+;  of SHA512 message blocks.
+;  "blocks" is the message length in SHA512 blocks
+; #######################################################################
+align 8
+global ASM_PFX(Sha512TransformAvx)
+ASM_PFX(Sha512TransformAvx):
+	test msglen, msglen
+	je nowork
+
+	; Allocate Stack Space
+	mov	rax, rsp
+	sub rsp, frame_size
+	and	rsp, ~(0x20 - 1)
+	mov	[rsp + frame_RSPSAVE], rax
+
+	; Save GPRs
+  ; Registers RBX, RBP, RDI, RSI, R12, R13, R14, R15, and XMM6-XMM15 are nonvolatile,
+  ; but XMM6-XMM15 are not used.
+	mov     [rsp + frame_GPRSAVE], rbx
+  mov     [rsp + frame_GPRSAVE + 8*1], rbp
```

Comment:
I think there is a mix of tabs and spaces in the asm code used for indentation, could you update to two spaces everywhere?